### PR TITLE
gradle2nix: init

### PR DESCRIPTION
**Pull request by ghost**
###### Motivation for this change

Based on #77417 
@tadfisher has done a wonderful job on [gradle2nix](https://github.com/tadfisher/gradle2nix) and this is an attempt to integrate it into nixpkgs.
I have chosen a strategy similar to the one used for yarn2nix: Only the required .nix files are copied from the upstream repository to nixpkgs, the nix code generator itself remains in a seperate repository and is packaged as a normal application.

Todo:

- [ ] reintroduce mxisd package tests as checkPhase
- [ ] debloat gradle2nix output
- [ ] change gradle2nix version to something other than "unspecified"
- ?

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @

### CODE REVIEW

**Comment by petabyteboy** [comment]
It seems like we are still having some IFD problems or something similar.

**Comment by volth** [comment]
Recursive Nix (https://github.com/NixOS/nix/pull/3205) might be a solution to avoid including big .json files into nixpkgs.
I am not sure, just an idea to investigate

**Comment by tadfisher** [comment]
Looks like everything built fine except `ma1sd`:

```
A problem occurred configuring root project 'source'.
> Could not resolve all artifacts for configuration ':classpath'.
   > Could not resolve com.github.ben-manes:gradle-versions-plugin:0.21.0.
     Required by:
         project :
      > No cached version of com.github.ben-manes:gradle-versions-plugin:0.21.0 available for offline mode.
```

This is strange because it depends on version `0.27.0` and `gradle2nix` resolved that correctly. I'll try and reproduce here.

**Review by fgaz — COMMENTED** [review]
Reviewed the changes to `shattered-pixel-dungeon`. I still have to actually build and test it.
edit: oops, I didn't notice the WIP status

File: pkgs/games/shattered-pixel-dungeon/default.nix [hunk 1]
```diff
@@ -20,50 +11,30 @@ let
...
-    outputHashMode = "recursive";
-    outputHash = "1k0v5scadw9ziq4dw2rckmh8x2xlmxslfsxmpw79zg78n3hvwhf1";
+  desktopItem = makeDesktopItem {
+    name = pname;
+    exec = "$out/bin/shattered-pixel-dungeon";
+    icon = "shattered-pixel-dungeon";
+    comment = meta.description;
+    desktopName = "Shattered Pixel Dungeon";
+    genericName = "Dungeon Game";
+    mimeType = "application/java;application/java-vm;application/java-archive";
```
[the remark this answers is not in the document]
Comment (fgaz) [reply pkgs/games/shattered-pixel-dungeon/default.nix]: Actually the mimeType property is for mime types the application can _handle_: https://specifications.freedesktop.org/desktop-entry-spec/desktop-entry-spec-latest.html#mime-types

So this line should be removed.

**Comment by tadfisher** [comment]
@petabyteboy Looks like `ma1sd` was checked out at the `2.2.1` branch instead of `2.1.1`.

Also, I have a concern with the approach here. Since the same `gradle-env.nix` is used for each package using `buildGradle`, I won't be able to change the JSON schema for `gradle-env.json`. I've already done so for gradle2nix 1.0.0-rc1, which would require an update to every package using `buildGradle`, both in `nixpkgs` and for out-of-tree callers importing `buildGradle` from  `nixpkgs`.

So there are a few options here:

1. Do nothing, and maybe leave a comment in the `gradle2nix` derivation that all `buildGradle` callers need to have `gradle2nix` re-run to generate a new gradle-env JSON file.
2. Make an `updateScript` function, similar to `gnome3.updateScript`, which calls `gradle2nix` when updating `buildGradle` packages, making it a bit more convenient to update `gradle-env.json` files all-at-once when updating `gradle2nix` itself.
3. Bundle `gradle-env.nix` with each package, instead of using the `yarn2nix`-style approach.
4. Modify `gradle2nix` to be a bit more convenient to bundle in `nixpkgs`. For example, flattening `gradle-env.json` into `gradle-env.nix` and making the latter a single file we can bundle in `nixpkgs`.

Thoughts?

**Comment by mguentner** [comment]
I packaged mxisd / ma1sd a while ago but stopped using both packages due to architectural changes in the matrix protocol (not relevant to this PR). This however is the reason why I am not paying much attention to the packages atm. I would be fine with removing at least mxisd soon and hope that somebody cares for ma1sd enough. Maybe some other user of ma1sd wants to take over!

**Comment by petabyteboy** [comment]
> @petabyteboy Looks like `ma1sd` was checked out at the `2.2.1` branch instead of `2.1.1`.
> 
> Also, I have a concern with the approach here. Since the same `gradle-env.nix` is used for each package using `buildGradle`, I won't be able to change the JSON schema for `gradle-env.json`. I've already done so for gradle2nix 1.0.0-rc1, which would require an update to every package using `buildGradle`, both in `nixpkgs` and for out-of-tree callers importing `buildGradle` from `nixpkgs`.

Thanks for noticing, I did not really think about this.
I think we should definitely get the changes done to the format in gradle2nix 1.0.0-rc1 into this PR. Of course it would be nice to see a more stable format, but if it does require another change, ...

> 
> So there are a few options here:
> 
> 1. Do nothing, and maybe leave a comment in the `gradle2nix` derivation that all `buildGradle` callers need to have `gradle2nix` re-run to generate a new gradle-env JSON file.

... I think option 1. is the best. since it is okay to require all gradle2nix applications to be updated when gradle2nix is has changes to the gradle-env.json format. Additionally I will add an updater script to each individual package to make this easier.

> 2. Make an `updateScript` function, similar to `gnome3.updateScript`, which calls `gradle2nix` when updating `buildGradle` packages, making it a bit more convenient to update `gradle-env.json` files all-at-once when updating `gradle2nix` itself.

I already wanted to implement an updater script for each package even without the possibility of changes to gradle-env.json, but now it is even more useful. 

> 3. Bundle `gradle-env.nix` with each package, instead of using the `yarn2nix`-style approach.

I am strongly opposed, since this would lead to code duplication, and in the end it would not be less work to do a proper upgrade of gradle2nix.

> 4. Modify `gradle2nix` to be a bit more convenient to bundle in `nixpkgs`. For example, flattening `gradle-env.json` into `gradle-env.nix` and making the latter a single file we can bundle in `nixpkgs`.

I am not sure what exactly you mean by this. I think the gradle2nix application itself is already easy to package. For other packages using buildGradle, I would see it as a big disadvantage if we have to duplicate the whole buildGradle expression for each package, so I am very happy with the current approach.

> 
> Thoughts?

File: pkgs/games/shattered-pixel-dungeon/default.nix [hunk 2]
```diff
@@ -1,15 +1,6 @@
...
-, fetchurl
-, makeWrapper
-, fetchFromGitHub
-, gradle_5
-, perl
-, jre
-, xorg
-, openal
-}:
+{ stdenv, fetchFromGitHub, buildGradle, jre, makeWrapper, xorg, openal, makeDesktopItem }:
```
[the remark this answers is not in the document]
Comment (petabyteboy) [reply pkgs/games/shattered-pixel-dungeon/default.nix]: Will do.

File: pkgs/games/shattered-pixel-dungeon/default.nix [hunk 3]
```diff
@@ -75,4 +46,3 @@ in stdenv.mkDerivation rec {
     platforms = platforms.all;
   };
 }
-
```
[the remark this answers is not in the document]
Comment (petabyteboy) [reply pkgs/games/shattered-pixel-dungeon/default.nix]: Will do.

File: pkgs/games/shattered-pixel-dungeon/default.nix (hunk 1 above)
[the remark this answers is not in the document]
Comment (petabyteboy) [reply pkgs/games/shattered-pixel-dungeon/default.nix]: Thanks, I accidentally copied it from the jd-gui package.

**Comment by tadfisher** [comment]
> ... I think option 1. is the best. since it is okay to require all gradle2nix applications to be updated when gradle2nix is has changes to the gradle-env.json format. Additionally I will add an updater script to each individual package to make this easier.

Sounds good!

There's also the possibility to include separate versions of gradle2nix if we need to make incompatible changes in the future, although I don't forsee this happening. Not a big deal for this PR, of course.

I wrote a skeleton `updateScript` wrapper expression, if helps, though I haven't tested it: https://gist.github.com/tadfisher/1007f68d321d6789dfd3b2301119f685

**Comment by petabyteboy** [comment]
Oh, that's neat. I already wrote updater-scripts for ma1sd and mxisd and fixed the NixOS test for them.
But it's not really duplicated work since your update script wouldn't work for them as it is, because they require the .git directory to be present when running gradle2nix, which is not present in the derivation source.

**Comment by petabyteboy** [comment]
I have noticed there is one big caveat to using update scripts:
When the gradle-env.json format changes, the update script must be called with the new gradle2nix env generator, but it must try to build it with the new gradle-env.nix or it will fail.
The script I wrote for mxisd and ma1sd work fine, because they don't try to evaluate the derivation.
I will think of something to avoid this, without giving up too much of the advantages that using nix-shell provides.

**Comment by tadfisher** [comment]
FYI, I pushed 1.0.0-rc2 which fixes a breaking behavior for Kotlin-based projects.

**Comment by Atemu** [comment]
Still important te me.

**Comment by petabyteboy** [comment]
I'm not planning to work on this anymore

**Pull request closed without merging**